Remove bail! errors from codebase

tyurek opened the pull request:
This PR removes all of the instances of bail! and bail_context! from the codebase. I introduced some new errors for cases which where previously handled by bails, but I'm starting to think that there are too many error variants to be useful. I like how FailedToVerifyProof consolidates different ways of a proof failing, and we could probably do something similar for some of these other variants.

I also think there's like three kinds of errors which would require different control flows to handle: 
"Expected": Things like looking for a contents from a message that hasn't been received yet. Eventually this will probably resolve and so not much more needs to be done besides trying again later
"Malicious": Things like a proof failing that should only happen if a Participant doesn't follow the protocol. Identifying and dealing with malicious behaviour will probably have its own control flow
"Bug": Things like InternalInvariantFailed which would only happen if there's a bug in the code.

tyurek commented on src/keygen/keygen_commit.rs:
```diff
@@ -93,7 +89,7 @@ impl KeygenDecommit {
         sender: &ParticipantIdentifier,
         com: &KeygenCommit,
-    ) -> Result<bool> {
+    ) -> Result<()> {
```
I changed this and other remaining functions verify functions to return errors instead of bools to be consistent with the rest of the codebase

gatoWololo replied:
Great! I like this better!

gatoWololo commented on src/errors.rs:
```diff
@@ -73,6 +93,15 @@ macro_rules! verify_err {
+    }};
+}
+
+#[allow(unused)]
```
If these are not used anymore we should just delete them. I'd love to eventually have no macros on this file!

gatoWololo commented on src/auxinfo/auxinfo_commit.rs:
```diff
@@ -95,6 +93,9 @@ impl AuxInfoDecommit {
         transcript.challenge_bytes(b"hashing r1", &mut hash);
         let rebuilt_com = AuxInfoCommit { hash };
-        Ok(rebuilt_com == *com)
+        match rebuilt_com == *com {
```
When matching like this on bools, we should just use an if statement!

gatoWololo commented on src/keygen/keygen_commit.rs:
```diff
@@ -102,6 +98,9 @@ impl KeygenDecommit {
         transcript.challenge_bytes(b"hashing r1", &mut hash);
         let rebuilt_com = KeygenCommit { hash };
-        Ok(rebuilt_com == *com)
+        match rebuilt_com == *com {
```
Same as previous comment. Use `if`.

gatoWololo commented on src/presign/round_one.rs:
```diff
@@ -72,7 +72,7 @@ impl Public {
             broadcasted_params,
         ) {
             Ok(()) => Ok(round_one_public),
-            Err(e) => bail!("Failed to verify round one public: {}", e),
+            Err(e) => Err(e),
```
This match can be further simplified as:
```rust
s can be further simplified as just:
```
round_one_public.verify(
&receiver_keygen_public.params,
&sender_keygen_public.pk,
broadcasted_params,
)?;
Ok(round_one_public)
```
```

gatoWololo commented on src/presign/round_three.rs:
```diff
@@ -81,7 +81,7 @@ impl Public {
         ) {
             Ok(()) => Ok(round_three_public),
-            Err(e) => bail!("Failed to verify round three public: {}", e),
+            Err(e) => Err(e),
```
This match can be simplfied like the other.

gatoWololo commented on src/presign/round_two.rs:
```diff
@@ -114,7 +114,7 @@ impl Public {
         ) {
             Ok(()) => Ok(round_two_public),
-            Err(e) => bail!("Failed to verify round two public: {}", e),
+            Err(e) => Err(e),
```
Simplify this one too.

gatoWololo commented on src/protocol.rs:
```diff
@@ -292,13 +292,11 @@ impl SignatureShare {
-        }?;
-
-        Ok(sig)
+        match self.r {
+            Some(r) => k256::ecdsa::Signature::from_scalars(r, s)
+                .map_err(|_| InternalError::SignatureInstantiationError),
+            None => Err(InternalError::NoChainedShares),
+        }
     }
```
We can get rid of the nesting here:
```
        let r = self.r.ok_or(InternalError::NoChainedShares)?;
        k256::ecdsa::Signature::from_scalars(r, s)
            .map_err(|_| InternalError::SignatureInstantiationError)
```

gatoWololo commented on src/auxinfo/auxinfo_commit.rs:
```diff
@@ -23,7 +23,7 @@ pub(crate) struct AuxInfoCommit {
     pub(crate) fn from_message(message: &Message) -> Result<Self> {
         if message.message_type() != MessageType::Auxinfo(AuxinfoMessageType::R1CommitHash) {
-            return bail!("Wrong message type, expected MessageType::Auxinfo(AuxinfoMessageType::R1CommitHash)");
+            return Err(crate::errors::InternalError::MisroutedMessage);
```
Instead of writing out the full crate path, we should import crate::errors::InternalError so we can just write `InternalError::MisroutedMessage` instead.

gatoWololo commented on src/keygen/participant.rs:
```diff
@@ -594,9 +592,7 @@ mod tests {
-            if participant.is_keygen_done(keygen_identifier).is_err() {
-                return bail!("Keygen not done");
-            }
+            participant.is_keygen_done(keygen_identifier)?;
```
Function names like `is_*` or `has_*` imply they return a bool. Since we are no longer returning a bool we should rename the function to something like `check_keygen_done`. Same for other functions you changed.

gatoWololo commented on src/presign/participant.rs:
```diff
@@ -470,14 +470,12 @@ impl PresignParticipant {
         // public auxinfo for other participants have been stored, since
         // this was a requirement to proceed for round 1.
-        if !has_collected_all_of_others(
+        has_collected_all_of_others(
```
Another fn which we should change the name. I like `check` instead of `has`.

gatoWololo commented on src/presign/participant.rs:
```diff
@@ -486,13 +484,15 @@ impl PresignParticipant {
             &self.storage,
             StorableType::PresignRoundTwoPrivate,
             message.id(),
-        )?;
+        )
+        .is_ok();
         let all_publics_received = has_collected_all_of_others(
             &self.other_participant_ids,
             &self.storage,
             StorableType::PresignRoundTwoPublic,
             message.id(),
-        )?;
+        )
+        .is_ok();
         if all_privates_received && all_publics_received {
```
You new code here does something slightly different from the original. In the original version we could return early with a Result<...> if the `?` encountered an error. The way the current code is written we never return early here. We always execute both `has_collected_all_of_others` checks and only return Ok(...).

I'm not sure if this is the intended behavior and if it is correct. Just wanted to point it out. What do you think?

tyurek replied:
I guess this distinction matters if `has_collected_all_of_others` can throw an error besides `StorageItemNotFound`. The only other possible error is in the `serialize!` macro which I don't think should be an issue. Some `StorageItemNotFound` errors are expected in general execution and shouldn't be propagated all the way to the top, so there has to be some code at some point that turns it into "not an error", right? 

Though this seems like a situation where distinguishing between expected errors and more serious ones will matter in the long term.

tyurek replied:
Maybe this is a reason to just have Storage-related queries return bools rather than errors?

gatoWololo replied:
> Maybe this is a reason to just have Storage-related queries return bools rather than errors? Then we can save errors for "something else went wrong" like unable to access storage or something

Right. That's the main disadvantage of having our errors mean something different in this context. Okay, to avoid having to think about what the error means for specific functions, I think we should change it so these functions return a `Result<bool, InternalError>` (I think that's the right type? This way we can always use `?` on functions and check the bool value directly.

In the future these functions may be able to return other types of 'unrecoverable` errors. (Maybe not but it is good to have our code be robust to changes like these).

tyurek replied:
I went ahead an changed functions like `contains_batch` and `is_keygen_done` to return `Result<bool, InternalError>`.

gatoWololo commented on src/presign/participant.rs:
```diff
@@ -836,15 +840,6 @@ impl PresignParticipant {
-            return bail!("Not ready to get other participants round three publics just yet!");
-        }
-
         let mut ret_vec = vec![];
```
How come the check here is no longer needed?

tyurek replied:
The idea was that if `has_collected_all_of_others()` fails then `storage.retrieve() ` will fail for one of the participants and you'll get the same result.

gatoWololo commented on src/errors.rs:
```diff
@@ -73,6 +93,15 @@ macro_rules! verify_err {
     }};
 }
 
+macro_rules! arg_err {
```
Functions are nicer than macros. I would remove this macro and rewrite it as a function:
```
impl InternalError {
    fn invalid_argument(error_message: &'static str) -> Self {
        InternalError::InvalidArgument(String::from(error_message))
    }
}
```

tyurek replied:
I styled this one after `verify_err!`, so whatever we do to one we should do to the other. Should I go ahead and swap both or save this for later?

gatoWololo replied:
Right, I saw there was other macros used in the code. I agree we should do it all in one go. We should save it for later then.

I will make a ticket for it, but it is not high priority.

gatoWololo replied:
This is now tracked on #152

gatoWololo commented on src/storage.rs:
```diff
@@ -153,33 +153,28 @@ impl Storage {
-                    storable_index
-                )
-            })?
+            .ok_or_else(|| InternalError::StorageItemNotFound(format!("{:?}", storable_index)))?
```
we could make a wrapper function for this error just like the other one. The signature would look like:
```rust
fn storage_item_not_found(d: impl Debug) -> InternalError {
    InternalError::StorageItemNotFound(format!("{:?}", d))
}
```

tyurek replied:
I actually just ended up simplifying the StorageItemNotFound error to not take a string, since in at least one instance it would have to print a fairly large string. If the additional information is needed, it's probably more appropriate to put it in a log

gatoWololo reviewed (changes requested):
I finished reviewing the PR. Overall I like the changes. I left some comments on little things to change and clean up.

I agree that we don't need so many error variants (and the list could just keep growing) and it is unclear if so much granularity is useful.

> I like how FailedToVerifyProof consolidates different ways of a proof failing, and we could probably do something similar for some of these other variants.

That sounds good to me. Would you like to do something similar as part of this PR? Or we can make a follow up PR too.

marsella commented:
> > I like how FailedToVerifyProof consolidates different ways of a proof failing, and we could probably do something similar for some of these other variants.
>
> That sounds good to me. Would you like to do something similar as part of this PR? Or we can make a follow up PR too.

Per discussion at the weekly meeting on 1/30, we'll make a separate follow-up issue to consolidate error variants and move some content to logs (pending #31).


Also, re: incredible slowness. I am now seeing 25-35 minute CI runtimes on other PRs as well, so that may not actually be an issue caused by this specifically.

tyurek commented:
Update: rebased against main and addressed most comments. There's one open question still and I'm leaving up the comments about semantic differences as I still need to test if they change performance

tyurek commented:
I wasn't able to recreate the slowness anymore, but I undid a semantic change that could make things slower anyway. Should be ready to go through once Omar and I work out the last minor details

gatoWololo commented:
@tomyurek Looks good!

> I wasn't able to recreate the slowness anymore (it could have been due to large variance in safe prime sampling times in the benchmark code, really), but I undid a semantic change that could make things slower anyway

This is fine. I will make a separate issue about general slowdowns and regression testing. Once we have some useful infrastructure for this, we can go back to past commits and narrow down slowdowns (if any pop up).

I think the last few things that are left:
- [ ] Address comment on using a `storage_item_not_found` function.
- [ ] Change functions to return bool instead of relying on Result error.
- [ ] Squash commits before merging!

gatoWololo reviewed (approved):
Looks good to me! Thanks for all the clean up!
